Table flush: wait until stopped

bhalevy opened the pull request:
When the table is stopped, all compaction groups
are stopped, and as part of that, they are flushing
their memtables.

table::flush can't enter the _pending_flushes_phaser
when the table is stopped since it is already closed,
so instead, just await wait_until_stopped() to wait
on any ongoing flushes that were induced by stop().

Fixes #22243

This is needed only since 4c89e624702fa71a1a29613b41890d7d0bc98b20
No backport required

tchaikov commented on replica/table.cc:
```diff
@@ -3057,6 +3057,9 @@ future<> table::flush(std::optional<db::replay_position> pos) {
         co_return;
     }
+    if (is_stopped()) {
+        co_return co_await wait_until_stopped();
```
> just await wait_until_stopped() to wait on any ongoing flushes that were induced by stop().

but why? why shall we wait until all current flush operations are completed. what could happen if we don't wait and just `co_return` here?

in the case of the fixed issue, the flush handlers are independent, so i don't think we have chained operations depending on the status of `table::flush()` 's completion.

bhalevy replied:
> > just await wait_until_stopped() to wait on any ongoing flushes that were induced by stop().
> 
> but why? why shall we wait until all current flush operations are completed. what could happen if we don't wait and just `co_return` here?

Then why do we wait in the regular path?
We can just trigger flush in the background and not wait for anything.
Since currently flush() may trigger memtable flush if needed, and or wait on ongoing flushes, I think it should behave the same also if ongoing flush happened to be triggered by table::stop().
That's part of its functional contract as I interpret it.

> 
> in the case of the fixed issue, the flush handlers are independent, so i don't think we have chained operations depending on the status of `table::flush()` 's completion.

Waiting is not for the commitlog caller, that indeed doesn't await the flush future.

denesb replied:
This seems like a layering violation to me: `flush()` "knows" that `stop()` will call flush, so `flush()` waits for `stop()`? Why cannot `stop()` just wait on the flushes itself?
If the problem is that `_pending_flushes_phaser.start()` below throws, why not check `_pending_flushes_phaser.closed()` and just return immediately if so?

bhalevy replied:
> This seems like a layering violation to me: 

But it's the same layer: class table

>`flush()` "knows" that `stop()` will call flush, so `flush()` waits for `stop()`? Why cannot `stop()` just wait on the flushes itself?

It does

> If the problem is that `_pending_flushes_phaser.start()` below throws, why not check `_pending_flushes_phaser.closed()` and just return immediately if so?

Because then it won't wait on ongoing flushes unlike the normal-operation mode contract.

elcallio replied:
Would it not make sense to simply treat it as an empty memtable, i.e. wait on `_flush_barrier` here (never closed), instead of adding another future?

bhalevy replied:
> Would it not make sense to simply treat it as an empty memtable, i.e. wait on `_flush_barrier` here (never closed), instead of adding another future?

This sounds like a good idea

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14502/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2821/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13950/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5942/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14502/)
#### Build Details:
- Duration: 4 hr 57 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
@denesb I concur with your comments, will send a simplified version.

bhalevy commented:
In v2 ([7feb7bc](https://github.com/scylladb/scylladb/commit/7feb7bceb2e01d496b85f52e859a1400106ba497)):
- used `std::optional<shared_future<>> _stopped` for waiting until stopped

bhalevy commented on replica/table.cc:
```diff
@@ -1700,6 +1700,7 @@ table::stop() {
         _sstables = make_compound_sstable_set();
     }));
     _cache.refresh_snapshot();
+    _stopped_promise.set_value();
```
Fixed in v2

bhalevy commented:
@denesb please re-review

elcallio commented on replica/database.hh:
```diff
@@ -939,6 +942,11 @@ public:
+        return _stopped.has_value();
+    }
+    // Must be called only after the table was stopped
+    future<> wait_until_stopped() const;
```
Should this not be private?

bhalevy replied:
Not sure why.  If stop() is public, why wouldn't wait_until_stopped() be public as well?

bhalevy replied:
Dropped in v3

elcallio commented:
Stop is public because of sharded object. `wait_until_stopped` here is wholly internal. 
Also, btw, calling `wait_until_stopped` in stop implies it would be acceptable to call this twice. Which it really is not. Sharded and all.

elcallio commented:
(And yes, I realize the "allow stop twice" is apparent in the earlier version as well - I am wondering slightly why that is).

bhalevy commented:
In v3 ([dfd3314](https://github.com/scylladb/scylladb/commit/dfd3314fb50a7c6bfacad095c0dd14a51dea2c61)):
- rewrote a much simplified fix:
  - table: flush: synchronize with stop()
  - Moved _pending_flushes_phaser.close() forward in table::stop, after stopping all storage/compaction groups
    - so that stop and flush can synchronize with each other safely

elcallio commented on replica/table.cc:
```diff
@@ -3057,6 +3062,10 @@ future<> table::flush(std::optional<db::replay_position> pos) {
         co_return;
     }
+    // There is nothing to flush if the table was stopped.
+    if (_pending_flushes_phaser.is_closed()) {
```
I am not quite getting this. Was not the intended effect earlier to _wait_ for any currently active flush to finish before we return here? Is the idea here that we've forced all flushes to disk through the call to `_sg_manager->stop_storage_groups();`? (The logic flow from there is somewhat indirect)?

bhalevy replied:
Yes, the change in table::stop to move the call to `_pending_flushes_phaser.close()` after `_sg_manager->stop_storage_groups()` means that is the phaser is_closed here, it means that all memtables were flushed via `stop_storage_groups` so there's nothing to wait for and we can just co_return immediately.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14590/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2893/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6012/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14021/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14590/)
#### Build Details:
- Duration: 4 hr 48 min
- Builder: spider6.cloudius-systems.com
